### PR TITLE
141 correct pagination@main

### LINKED ISSUE #141: key columns missing in pages other than first page if table is paginated in columns
if we do `export_as_txt` on rlisting object, the key columns will only be displayed in the first page instead of every page if paginated by columns.

example:

```R
tbl <- as_listing(
  iris,
  key_cols = c("Species"),
  disp_cols = c("Sepal.Length", "Sepal.Width", "Petal.Length", "Petal.Width")
)

export_as_txt(
  head(tbl),
  cpp = 50,
  page_break= "\f"
) %>%
  cat()

```

we will see
page 1
```
Species   Sepal.Length   Sepal.Width
————————————————————————————————————
setosa        5.1            3.5    
              4.9             3     
              4.7            3.2    
              4.6            3.1    
               5             3.6    
              5.4            3.9    
```
page 2
```
Petal.Length   Petal.Width
——————————————————————————
    1.4            0.2    
    1.4            0.2    
    1.3            0.2    
    1.5            0.2    
    1.4            0.2    
    1.7            0.4
```

apparently, in page 2, without the key column displayed, it is difficult to read the information

raised by @w316liu 

update: there are other issues in pagination

```R
export_as_txt(
  head(tbl),
  cpp = 30,
  page_break= "\f"
) %>%
  cat()
```
will give error like 

```
Error in .chkdim_and_replace(mf, value, component = "spans") : 
  Dimensions of new 'spans' value (1, 7) do not match dimensions of existing 'strings' component (7, 1).
```
session info
```
R version 4.2.2 Patched (2022-11-10 r83330)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] rlistings_0.2.3.9000  tibble_3.1.8          formatters_0.5.1.9000

loaded via a namespace (and not attached):
 [1] fansi_1.0.4     digest_0.6.31   utf8_1.2.3      grid_4.2.2     
 [5] lifecycle_1.0.3 jsonlite_1.8.4  backports_1.4.1 magrittr_2.0.3 
 [9] pillar_1.8.1    rlang_1.0.6     cli_3.6.0       vctrs_0.5.2    
[13] checkmate_2.1.0 glue_1.6.2      fastmap_1.1.0   compiler_4.2.2 
[17] pkgconfig_2.0.3 htmltools_0.5.4

```

### PR DESCRIPTION
**Pull request by ayogasekaram**
closes #141 

related to https://github.com/insightsengineering/formatters/pull/186

1. merge formatters change
2. Update description with new formatters version.

### CODE REVIEW

File: R/listing_export.R [hunk 1]
```diff
@@ -10,6 +10,6 @@
 #'   add_listing_col("AESER / AREL", fun = function(df) paste(df$AESER, df$AREL, sep = " / "))
 #' main_title(lsting) <- "this is some title"
 #' main_footer(lsting) <- "this is some footer"
-#' cat(export_as_txt(lsting, file = NULL, paginate = TRUE))
+#' cat(export_as_txt(lsting, file = NULL, paginate = TRUE, rep_cols = length(get_keycols(lsting))))
```
Comment (Melkiades) [inline R/listing_export.R:13]: Should be this automatic?

File: tests/testthat/_snaps/export.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# key columns repeat with export_as_txt
+
+    Code
+      listing_exp
+    Output
+      [1] "Unique Subject Identifier   Age   Continous Level Biomarker 1\n—————————————————————————————————————————————————————————————\n  AB12345-BRA-1-id-134      47    6.5                        \n  AB12345-BRA-1-id-141      35    7.5                        \n\\s\\nUnique Subject Identifier   Age   Continous Level Biomarker 1\n—————————————————————————————————————————————————————————————\n  AB12345-BRA-1-id-236      32    7.7                        \n  AB12345-BRA-1-id-265      25    10.3                       \n\\s\\nUnique Subject Identifier   Age   Continous Level Biomarker 1\n—————————————————————————————————————————————————————————————\n   AB12345-BRA-1-id-42      36    2.3                        \n   AB12345-BRA-1-id-65      25    7.3                        \n"
```
Comment (Melkiades) [inline tests/testthat/_snaps/export.md:6]: can you capture it as `cat`? This is not easy to read

File: R/listing_export.R (hunk 1 above)
Comment (ayogasekaram) [reply R/listing_export.R:13]: I thought of changing this in the export_as_txt function but right now it uses rep_cols=num_cols. The method doesn't look like it's capturing the key_cols correctly.

**Review by shajoezhu — CHANGES_REQUESTED** [review]
hi @ayogasekaram , thanks for the changes, can you also include Liming's example in the test as well. 

https://github.com/insightsengineering/rlistings/issues/141#issue-1826731597

The main issue here is that horizontal pagination drops the key columns. I don't think the current test case is sufficient. Thanks

File: R/listing_export.R (hunk 1 above)
Comment (Melkiades) [reply R/listing_export.R:13]: I think we may need to change that. I do not know. @shajoezhu @gmbecker ?

File: tests/testthat/test-export.R [hunk 3]
```diff
@@ -7,16 +7,47 @@ testthat::test_that("Listing print correctly", {
...
+  # pre-processing and ordering
+  tbl <- as_listing(
+    iris,
+    key_cols = c("Species"),
+    disp_cols = c("Sepal.Length", "Sepal.Width", "Petal.Length", "Petal.Width")
+  )
+
+  listing <- suppressMessages(export_as_txt(tbl, lpp = 20, paginate = TRUE,
+                                            page_break = "\n",
+                                            rep_cols = length(get_keycols(tbl))))
```
Comment (Melkiades) [inline tests/testthat/test-export.R:39]: I think it is better to strsplit on the original page_break and reduce the output to be snapshot

File: tests/testthat/test-export.R [hunk 4]
```diff
@@ -7,16 +7,47 @@ testthat::test_that("Listing print correctly", {
...
+    dplyr::distinct(USUBJID, AGE, BMRKR1, .keep_all = TRUE)
+
+  lsting <- as_listing(tmp_data,
+                       key_cols = c("USUBJID", "AGE"),
+                       disp_cols = character()
+  ) %>%
+    add_listing_col("BMRKR1", format = "xx.x")
+
+  listing_exp <- suppressMessages(export_as_txt(lsting, lpp = 4, verbose = TRUE, page_break = "\n",
+                                                rep_cols = length(get_keycols(lsting))))
```
Comment (Melkiades) [inline tests/testthat/test-export.R:23]: @shajoezhu I think the second direct regression test is not needed. This is covering the same bug, I would add as comment the number of the issue

File: tests/testthat/_snaps/export.md [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+                       4.4            2.9           1.4            0.2    
+                       4.9            3.1           1.5            0.1    
+                       5.4            3.7           1.5            0.2    
+                       4.8            3.4           1.6            0.2    
+                       4.8             3            1.4            0.1    
+                       4.3             3            1.1            0.1    
+                       5.8             4            1.2            0.2    
+                       5.7            4.4           1.5            0.4    
+                       5.4            3.9           1.3            0.4    
+                       5.1            3.5           1.4            0.3
```
Comment (shajoezhu) [inline tests/testthat/_snaps/export.md:45]: hi @ayogasekaram and @Melkiades , thanks for looking into this again. but the test is still need work

We would like to see horizontal pagination tested here. i.e. the snapshot should like 

image

File: R/listing_export.R (hunk 1 above)
Comment (Melkiades) [inline R/listing_export.R:13]: do we need this still?

**Review by Melkiades — APPROVED** [review]
The solution with the generic is very good! Remember only to up the `formatters` version in the description and remove from the example the rep_cols (or change it to another custom value, to show the feature). Thanks Abi :)

File: R/listing_export.R (hunk 1 above)
Comment (shajoezhu) [reply R/listing_export.R:13]: yes, if not, the horizontal pagination will not work.

**Review by shajoezhu — APPROVED** [review]
lgtm! thanks @ayogasekaram

**Comment by shajoezhu** [comment]
very nice! thanks @Melkiades @ayogasekaram

**Pull request merged**